# Probability

## Overall Shooting Statistics

On this slide, calculate and report the:
- Overall probability of a make (result = TRUE)
- Overall probability of a miss (result = FALSE)
- Overall proportion of three-pointers (shot_type = 3)
- Overall proportion of two-pointers (shot_type = 2)

|  | calculation |
| :--- | :--- |
| $P(makesShot)$ | $\frac{\normalsize numMade}{\normalsize numShots} = \frac{\normalsize 703}{\normalsize 1434}\approx0.4902$ |
| $P(\neg{makesShot})$ | $1 - P(makesShot) = 1 - 0.4902370990237099\approx0.5098$ |
| $P(threePointer)$ | $\frac{\normalsize numThreePointers}{\normalsize numShots} = \frac{\normalsize 786}{\normalsize 1434}\approx0.5481$ |
| $P(twoPointer)$ | $\frac{\normalsize numTwoPointers}{\normalsize numShots} = \frac{\normalsize 648 }{\normalsize 1434}$ or $1 - P(threePointer) \approx0.4519$|


In [1]:
import numpy as np
import pandas as pd
from scipy.stats import binom

shot_data = pd.read_csv('./data/stephen_curry_shots_2023.csv')

shot_results = shot_data.value_counts('result')
nShots = shot_data.index.size
nShotsMade = shot_results[True]

pMake = nShotsMade / nShots

shot_type = shot_data.value_counts('shot_type')
nThreePointer = shot_type[3]
pThreePointer = nThreePointer / nShots
nTwoPointer = shot_type[2]
pTwoPointer = nTwoPointer / nShots

print(f'P(makesShot) = {pMake},{pMake:.4g}')
print(f'P(notMakesShot) = {1 - pMake}, {1 - pMake:.4g}')
print(f'P(threePointer) = {pThreePointer}, {pThreePointer:.4g}')
print(f'P(twoPointer) = {pTwoPointer}, {pTwoPointer:.4g}')

P(makesShot) = 0.4902370990237099,0.4902
P(notMakesShot) = 0.5097629009762901, 0.5098
P(threePointer) = 0.5481171548117155, 0.5481
P(twoPointer) = 0.45188284518828453, 0.4519


## Probabilties

Given the results you found on the previous slide:

- What is the probability of Steph making 3 of the next 4 shots?
- What is the probability that 4 of the next 5 shots are three-pointers?

| | calculation | result |
| :--- | :--- | :--- |
| prob of making 3 of next 4 shots | `binom.pmf(3, 4, P(makesShot))` | 24.02% |
| prob of 4 of 5 next shots being 3 pointers | `binom.pmf(4, 5, P(threePointer))` | 20.39% |

### Assumptions
In these calculations we are considering all shots made across the entire 2023 season as independent events. The means we are assuming shot probabilities do not change over time (indicated by date or game quarter) or influenced by other factors.

In [2]:
pMake3of4 = binom.pmf(3, 4, pMake)
p4of5ThreePointers = binom.pmf(4, 5, pThreePointer)

print(f'3 of 4 made shots probability (any possible combination): {pMake3of4}, {pMake3of4:.2%}')
print(f'4 of 5 three pointers probability (any possible combination): {p4of5ThreePointers}, {p4of5ThreePointers:.2%}')

3 of 4 made shots probability (any possible combination): 0.2402407848582661, 24.02%
4 of 5 three pointers probability (any possible combination): 0.20393390610921378, 20.39%


## Conditional Probabilities - Future
Contrary to the instructions, Bayes theorem is not required for an accurate calculation, but has been demonstrated below along with simplier method

$P(makesShot \mid threePointer) = \frac{\normalsize P(threePointer \mid makesShot) \times P(makesShot)}{\normalsize P(threePointer)} = \frac{\normalsize 0.46799431009957326 \times 0.4902370990237099}{\normalsize 0.5481171548117155} \approx 0.418575$

$P(makesShot \mid threePointer) = \frac{\normalsize numThreePointerWhileLead}{\normalsize numShotsWhileLead} = \frac{\normalsize 388}{\normalsize 722} \approx 0.418575$

$P(lead \mid threePointer) = \frac{\normalsize numThreePointerWhileLead}{\normalsize numThreePointer} = \frac{\normalsize 388}{\normalsize 786} = 0.49363867684478374 \approx 0.4936$

$P(threePointer \mid lead) = \frac{\normalsize P(lead \mid threePointer) \times P(threePointer)}{\normalsize P(lead)} = \frac{\normalsize 0.49363867684478374 \times 0.5481171548117155}{\normalsize 0.5034867503486751} = 0.5373961218836566 \approx 0.537396$

$P(threePointer \mid lead) = \frac{\normalsize numThreePointerWhileLead}{\normalsize numShotsWhileLead} = \frac{\normalsize 388}{\normalsize 722} = 0.5373961218836565 \approx 0.537396$



$P(makesShot \mid twoPointer) = \frac{\normalsize P(twoPointer \mid makesShot) \times P(makesShot)}{\normalsize P(twoPointer)} = \frac{\normalsize 0.5320056899004267 \times 0.4902370990237099}{\normalsize 0.45188284518828453}$

### Q: If the next shot Steph shoots is a three-pointer, what is the probability he makes it?
### A:
$P(makesShot \mid threePointer)\approx0.4186$

#### calculation
```python
import pandas as pd

shot_data = pd.read_csv('./data/stephen_curry_shots_2023.csv')

threePointShots = shot_data[shot_data.shot_type == 3]
nThreePointShots = threePointShots.index.size # 786
nMade_threePointShots = threePointShots[threePointShots.result == True].index.size # 329
pMadeIfThreePointer = nMade_threePointShots / nThreePointShots # 0.41857506361323155
```

### Q: what is the probability it was taken while his team had the lead
### A:
$P(lead \mid threePointer)\approx0.4936$

#### calculation
```python
nThreePointShotsWhileLead = threePointShots[threePointShots.lead == True].index.size # 388
pLeadIfThreePointer = nThreePointShotsWhileLead / nThreePointShots # 0.49363867684478374
```
---

### Q: If the next shot Steph shoots is a two-pointer, what is the probability he makes it?
### A:
$P(makesShot \mid twoPointer)\approx0.5772$

#### calculation
```python
twoPointShots = shot_data[shot_data.shot_type == 2]
nTwoPointShots = twoPointShots.index.size # 648
nMade_twoPointShots = twoPointShots[twoPointShots.result == True].index.size # 374
pMadeIfTwoPointer = nMade_twoPointShots / nTwoPointShots # 0.5771604938271605
```

### Q: what is the probability it was taken while his team had the lead
### A:
$P(lead \mid twoPointer)\approx0.5154$

#### calculation
```python
nTwoPointShotsWhileLead = twoPointShots[twoPointShots.lead == True].index.size # 334
pLeadIfTwoPointer = nTwoPointShotsWhileLead / nTwoPointShots # 0.5154320987654321
```

In [ ]:
print('\n3-pointer\nwhat is the probability he makes it?')

threePointShots = shot_data[shot_data.shot_type == 3]
nMade_threePointShots = threePointShots[threePointShots.result == True].index.size

# Bayes
pThreePointerIfMade = nMade_threePointShots/nShotsMade
pMadeIfThreePointerBayes = (pThreePointerIfMade * pMake)/pThreePointer
print(f'Bayes: P(makesShot | ThreePointer) = {pMadeIfThreePointerBayes}, {pMadeIfThreePointerBayes:.2%}')

# Simple
pMadeIfThreePointerSimple = nMade_threePointShots / threePointShots.index.size
print(f'Simple: P(makesShot | ThreePointer) = {pMadeIfThreePointerSimple}, {pMadeIfThreePointerSimple:.2%}')



print('\nwhat is the probability it was taken while his team had the lead')

leadShots = shot_data[shot_data.lead == True]
nShotsWhileLead = leadShots.index.size # 722
nThreePointerWhileLead = leadShots[leadShots.shot_type == 3].index.size # 388
pLead = nShotsWhileLead / nShots # 0.5034867503486751

# Bayes
pLeadWhenThreePointer = nThreePointerWhileLead / nThreePointer
pThreePointerWhenLeadBayes = (pLeadWhenThreePointer * pThreePointer) / pLead
print(f'Bayes: p(threePointer | lead) = {pThreePointerWhenLeadBayes}, {pThreePointerWhenLeadBayes:.2%}')

#Simple
pThreePointerWhenLeadSimple = nThreePointerWhileLead / nShotsWhileLead
print(f'Simple: P(threePointer | lead) = {pThreePointerWhenLeadSimple}, {pThreePointerWhenLeadSimple:.2%}')


print('\n2-pointer\nwhat is the probability he makes it?')
# Simple
twoPointShots = shot_data[shot_data.shot_type == 2]
nMade_twoPointShots = twoPointShots[twoPointShots.result == True].index.size # 374
pMadeIfTwoPointerSimple = nMade_twoPointShots / twoPointShots.index.size
print(f'Simple: P(makesShot | TwoPointer) = {pMadeIfTwoPointerSimple}, {pMadeIfTwoPointerSimple:.2%}')

# Bayes
pTwoPointerIfMade = nMade_twoPointShots/nShotsMade # 0.5320056899004267
pMadeIfTwoPointerBayes = (pTwoPointerIfMade * pMake)/pTwoPointer
print(f'Bayes: P(makesShot | twoPointer) = {pMadeIfTwoPointerBayes}, {pMadeIfTwoPointerBayes:.2%}')

print('\nwhat is the probability it was taken while his team had the lead')
nTwoPointerWhileLead = leadShots[leadShots.shot_type == 2].index.size # 334
#Simple
pTwoPointerWhenLeadSimple = nTwoPointerWhileLead / nShotsWhileLead
print(f'Simple: P(twoPointer | lead) = {pTwoPointerWhenLeadSimple}, {pTwoPointerWhenLeadSimple:.2%}')

# Bayes
pLeadWhenTwoPointer = nTwoPointerWhileLead / nTwoPointer # 0.5154320987654321
pTwoPointerWhenLeadBayes = (pLeadWhenTwoPointer * pTwoPointer) / pLead
print(f'Bayes: P(twoPointer | lead) = {pTwoPointerWhenLeadBayes}, {pTwoPointerWhenLeadBayes:.2%}')



3-pointer
what is the probability he makes it?
Simple: P(makesShot | ThreePointer) = 0.41857506361323155, 41.86%
Bayes: P(makesShot | ThreePointer) = 0.4185750636132315, 41.86%

what is the probability it was taken while his team had the lead
Simple: P(threePointer | lead) = 0.5373961218836565, 53.74%
Bayes: p(threePointer | lead) = 0.5373961218836566, 53.74%

2-pointer
what is the probability he makes it?
Simple: P(makesShot | TwoPointer) = 0.5771604938271605, 57.72%
Bayes: P(makesShot | twoPointer) = 0.5771604938271604, 57.72%

what is the probability it was taken while his team had the lead
Simple: P(twoPointer | lead) = 0.4626038781163435, 46.26%
Bayes: P(twoPointer | lead) = 0.46260387811634346, 46.26%


## Conditional Probabilities - Past

### Q: If Steph just made a shot, what is the probability that it was a three-pointer

### A: 0.468

|  |  |
| :--- | :--- |
| P(makesShot) | 0.4902 |
| P(threePointer) | 0.5481 |
| P(makesShot\|threePointer) | 0.4186 |

$P(threePointer \mid makesShot) = \frac{\normalsize P(makesShot|threePointer) \times P(threePointer)}{\normalsize P(makesShot)}$

$P(threePointer \mid makesShot) = \frac{\normalsize 0.41857506361323155 \times 0.5481171548117155}{\normalsize 0.4902370990237099} = 0.4679943100995733 \approx 0.468$

### Q: What is the probability it was a two-pointer

### A: 0.532

$P(makesShot) = 0.4902370990237099$

$P(twoPointer) = 0.45188284518828453$

$P(makesShot \mid twoPointer) = 0.5771604938271605$

$P(twoPointer \mid makesShot) = \frac{\normalsize P(makesShot|twoPointer) \times P(twoPointer)}{\normalsize P(makesShot)}$

$P(twoPointer \mid makesShot) = \frac{\normalsize 0.5771604938271605 \times 0.45188284518828453}{\normalsize 0.4902370990237099} = 0.5320056899004268\approx0.532$

In [4]:
print(f'P(makesShot) = {pMake:.4g}')
print(f'P(threePointer) = {pThreePointer:.4g}')
print(f'P(makesShot|threePointer) = {pMadeIfThreePointer:.4g}')
print(f'P(threePointer|makesShot) = {pThreePointerIfMade:.4g}')
print(f'P(twoPointer) = {pTwoPointer:.4g}')
print(f'P(makesShot|twoPointer) = {pMadeIfTwoPointer:.4g}')
print(f'P(twoPointer|makesShot) = {pTwoPointerIfMade:.4g}')

P(makesShot) = 0.4902
P(threePointer) = 0.5481


NameError: name 'pMadeIfThreePointer' is not defined